src/core/ext/filters/stateful_session_affinity/stateful_session_affinity_filter.cc

```diff
@@ -0,0 +1,231 @@
+  auto cookie_value = GetHostOverrideFromCookie(
+      call_args.client_initial_metadata, *cookie_config->name);
+  if (cookie_value.has_value()) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_stateful_session_affinity_filter_trace)) {
+      gpr_log(GPR_INFO,
+              "chand=%p: stateful session affinity found cookie %s value %s",
+              this, cookie_config->name->c_str(),
+              std::string(*cookie_value).c_str());
+    }
+    // We have a valid cookie, so add the call attribute to be used by the
+    // xds_override_host LB policy.
+    service_config_call_data->SetCallAttribute(XdsHostOverrideTypeName(),
+                                               *cookie_value);
+  }
+  // Intercept server initial metadata.
+  auto* read_latch = GetContext<Arena>()->New<Latch<ServerMetadata*>>();
+  auto* write_latch =
+      std::exchange(call_args.server_initial_metadata, read_latch);
+  return TryConcurrently(
+             Seq(next_promise_factory(std::move(call_args)),
+                 [cookie_config, cookie_value](ServerMetadataHandle md) {
+                   MaybeUpdateServerInitialMetadata(cookie_config, cookie_value,
+                                                    md.get());
+                   return md;
+                 }))
+      .NecessaryPull(Seq(read_latch->Wait(),
+                         [write_latch, cookie_config,
+                          cookie_value](ServerMetadata** md) -> absl::Status {
+                           if (*md != nullptr) {
+                             MaybeUpdateServerInitialMetadata(
+                                 cookie_config, cookie_value, *md);
+                           }
+                           write_latch->Set(*md);
+                           return absl::OkStatus();
+                         }));
+}
+
+absl::optional<absl::string_view>
+StatefulSessionAffinityFilter::GetHostOverrideFromCookie(
+    const ClientMetadataHandle& client_initial_metadata,
+    absl::string_view cookie_name) {
+  // Check to see if the cookie header is present.
+  std::string buffer;
+  auto header_value =
+      client_initial_metadata->GetStringValue("cookie", &buffer);
+  if (!header_value.has_value()) return absl::nullopt;
+  // Parse cookie header.
+  std::vector<absl::string_view> values;
+  for (absl::string_view cookie : absl::StrSplit(*header_value, "; ")) {
+    std::pair<absl::string_view, absl::string_view> kv =
+        absl::StrSplit(cookie, absl::MaxSplits('=', 1));
+    if (kv.first == cookie_name) values.push_back(kv.second);
+  }
+  if (values.empty()) return absl::nullopt;
+  // TODO(roth): Figure out the right behavior for multiple cookies.
+  // For now, just choose the first value.
+  absl::string_view value = values.front();
+  // Base64-decode it.
+  std::string decoded_value;
+  if (!absl::Base64Unescape(value, &decoded_value)) return absl::nullopt;
+  // Copy it into the arena, since it will need to persist until the LB pick.
```

Comment:
Why not just return an std::string? I understand it may cause some unnecessary copies but compilers are good with stuff like return value optimization and inlining...

Reply:
Replied in the other PR.